doc/build/snippets/writing.rst

```diff
@@ -0,0 +1,162 @@
+Writing Snippets
+################
+
+.. contents::
+   :local:
+
+Basics
+******
+
+Snippets are defined using YAML files named :file:`snippet.yml`.
+
+A :file:`snippet.yml` file contains the name of the snippet, along with
+additional build system settings, like this:
+
+.. code-block:: yaml
+
+   name: snippet-name
+   # ... build system settings go here ...
+
+Build system settings go in other keys in the file as described later on in
+this page.
+
+You can combine settings whenever they appear under the same keys. For example,
+you can combine a snippet-specific devicetree overlay and Kconfig fragment like
+this:
+
+.. code-block:: yaml
+
+   name: foo
+   append:
+     DTC_OVERLAY_FILE: foo.overlay
+     OVERLAY_CONFIG: foo.conf
+
+Where snippets are located
+**************************
+
+The build system looks for snippets in these places:
+
+#. In directories configured by the :makevar:`SNIPPET_ROOT` CMake variable.
+   This always includes the zephyr repository (so
+   :zephyr_file:`zephyr/snippets` is always a source of snippets) and the
+   application source directory (so :file:`<app>/snippets` is also).
+
+   Additional directories can be added manually at CMake time.
+
+   The variable is a whitespace- or semicolon-separated list of directories
+   which may contain snippet definitions.
+
+   For each directory in the list, the build system looks for
+   :file:`snippet.yml` files underneath a subdirectory named :file:`snippets/`,
+   if one exists.
+
+   For example, if :makevar:`SNIPPET_ROOT` is set to ``/foo;/bar``, the build
+   system will look for :file:`snippet.yml` files underneath the following
+   subdirectories:
+
+   - :file:`/foo/snippets/`
+   - :file:`/bar/snippets/`
+
+   The :file:`snippet.yml` files can be nested anywhere underneath these
+   locations.
+
+#. In any :ref:`module <modules>` whose :file:`module.yml` file provides a
+   ``snippet_root`` setting.
+
+   For example, in a zephyr module named ``baz``, you can add this to your
+   :file:`module.yml` file:
+
+   .. code-block:: yaml
+
+      settings:
+        snippet_root: .
+
+   And then any :file:`snippet.yml` files in ``baz/snippets`` will
+   automatically be discovered by the build system, just as if
+   the path to ``baz`` had appeared in :makevar:`SNIPPET_ROOT`.
+
+Processing order
+****************
+
+The order that snippets are processed is currently not defined.
+Therefore, you should write your :file:`snippet.yml` file so that
+it is not dependent on other snippets.
```

Comment:
I believe that the processing order, even if we assert that snippets should not be co-dependent, is still an important aspect, and we should clearly define it now to improve the usability of the snippets feature.

One use case I can think of is, for a USB device, the `CONFIG_USB_DEVICE_PRODUCT` is initially set to `Zephyr USB console sample` by the `cdc-acm-console` snippet.

Now if you want to have another set of snippets that control the "branding" of your product (e.g. `brand-nordic`, `brand-intel`, ...) and override `CONFIG_USB_DEVICE_PRODUCT` as well as other product information-related configurations, it is important that the Kconfig overlay from the `brand-nordic` snippet is applied after that of the `cdc-acm-console` snippet.

I wonder if we should define the processing order of the snippets to be "in the order they are listed in the `SNIPPET` variable so as to allow the use cases like the above.